tests/test-update-refspec-endoflife.c

```diff
@@ -0,0 +1,243 @@
+
+  eos_test_client_reap_updater (client,
+                                &updater_cmd,
+                                &reaped,
+                                &error);
+  g_assert_no_error (error);
+
+  cmds = g_ptr_array_new ();
+  g_ptr_array_add (cmds, &reaped);
+  g_ptr_array_add (cmds, autoupdater->cmd);
+  g_assert_true (cmd_result_ensure_all_ok_verbose (cmds));
+}
+
+/* Start with a commit, and then make a final commit on the first refspec which
+ * adds a redirect, which should change the upgrade ref. */
+static void
+test_update_refspec_endoflife (EosUpdaterFixture *fixture,
+                               gconstpointer user_data)
+{
+  g_autoptr(GFile) server_root = NULL;
+  g_autoptr(EosTestServer) server = NULL;
+  g_autofree gchar *keyid = get_keyid (fixture->gpg_home);
+  g_autoptr(GError) error = NULL;
+  g_autoptr(EosTestSubserver) subserver = NULL;
+  g_autoptr(GFile) client_root = NULL;
+  g_autoptr(EosTestClient) client = NULL;
+  g_autoptr(GFile) repo_path = NULL;
+  g_autoptr(OstreeRepo) repo = NULL;
+  g_autoptr(GHashTable) additional_metadata_for_commit = NULL;
+  g_autoptr(GHashTable) leaf_commit_nodes =
+    eos_test_subserver_ref_to_commit_new ();
+  gboolean has_commit;
+  g_autofree gchar *branches_option = NULL;
+  g_autofree gchar *expected_branches = NULL;
+
+  /* We could get OSTree working by setting OSTREE_BOOTID, but shortly
+   * afterwards we hit unsupported syscalls in qemu-user when running in an
+   * ARM chroot (for example), so just bail. */
+  if (!eos_test_has_ostree_boot_id ())
+    {
+      g_test_skip ("OSTree will not work without a boot ID");
+      return;
+    }
+
+  insert_update_refspec_metadata_for_commit (1,
+                                             next_ref,
+                                             &additional_metadata_for_commit);
+
+  server_root = g_file_get_child (fixture->tmpdir, "main");
+  server = eos_test_server_new_quick (server_root,
+                                      default_vendor,
+                                      default_product,
+                                      default_collection_ref,
+                                      0,
+                                      fixture->gpg_home,
+                                      keyid,
+                                      default_ostree_path,
+                                      NULL,
+                                      NULL,
+                                      additional_metadata_for_commit,
+                                      &error);
+  g_assert_no_error (error);
+  g_assert_cmpuint (server->subservers->len, ==, 1u);
+
+  subserver = g_object_ref (EOS_TEST_SUBSERVER (g_ptr_array_index (server->subservers, 0)));
+  client_root = g_file_get_child (fixture->tmpdir, "client");
+  client = eos_test_client_new (client_root,
+                                default_remote_name,
+                                subserver,
+                                default_collection_ref,
```

Comment:
Can you also please do a version of this test without collection IDs in the configuration? I don’t think we currently have any coverage for `ostree.endoflife-rebase`.

Reply:
done (on the client only because the server will forever have a collection ID set)